### ISSUE TITLE
{Far Future] ModAchievements & tML Achievements

### ISSUE THREAD

**Issue by Solxanich** [tModLoader/tModLoader#1491]
### Description
Add a new system for having in-game achievements in tModLoader, that are independent to steam.
An example reference for ModAchievements could be the modded library [Achievement Lib](https://mirror.sgkoi.dev/Mods/Details/AchievementLib)

It may also be fruitful to consider tML specific achievements (such as playing X hours, using more than 4GB of memory, wearing a full patreon/dev set, playing ExampleMod bc example achievement) 

### What does this proposal attempt to solve or improve?
Reduction of 3rd party libraries that accomplish features that should be in tML.

### Which (other) solutions should be considered?

TBD, this is FAR future

**Comment by Patcheresu** [comment]
Some suggestions:
Playing TML for a certain number of hours
Creating a world that has a mod alter worldgen
Downloading 'any' mod
Making 'any' mod
Assigning a key to 'any' mod altered hotkey in the settings menu
Beating a Pre-Hardmode non-Vanilla boss in TML
Beating a Hardmode non-Vanilla boss in TML
Making furniture, finding weapons, crafting items that aren't from Vanilla
Beating Moon Lord in TML

**Comment by Steviegt6** [comment]
> Some suggestions:
> Playing TML for a certain number of hours
> Creating a world that has a mod alter worldgen
> Downloading 'any' mod
> Making 'any' mod
> Assigning a key to 'any' mod altered hotkey in the settings menu
> Beating a Pre-Hardmode non-Vanilla boss in TML
> Beating a Hardmode non-Vanilla boss in TML
> Making furniture, finding weapons, crafting items that aren't from Vanilla
> Beating Moon Lord in TML

Pre-made tML achievements are kind of dumb. tML is a modding API and mod loader, not a content mod in itself (save for some Patreon sets, an old April Fool's item back from when Blushie was the main dev, and a few items to help with mod compatibility (unloaded items, Starter Bag).

**Comment by Steviegt6** [comment]
> ### Description
> Add a new system for having in-game achievements in tModLoader, that are independent to steam.
> An example reference for ModAchievements could be the modded library [Achievement Lib](https://mirror.sgkoi.dev/Mods/Details/AchievementLib)
> 
> It may also be fruitful to consider tML specific achievements (such as playing X hours, using more than 4GB of memory, wearing a full patreon/dev set, playing ExampleMod bc example achievement, setting a hotkey, downloading any mod)
> 
> ### What does this proposal attempt to solve or improve?
> Reduction of 3rd party libraries that accomplish features that should be in tML.
> 
> ### Which (other) solutions should be considered?
> TBD, this is FAR future

Is there any current reason this is considered something that would be added far in the future? Terraria already has the framework for getting this all to work, tML would essentially just need to wrap around the current system.

A big point of discussion would be whether or not we keep vanilla achievements, and how we'll save modded achievements. I'm not sure of how how we'd properly identify modded achievements (we could save each mod in their own file, but I'm not sure how nice that'd be. We could identify with string lexers and save all modded data to a JSON or some other file format, I suppose). I believe we've discussed this before but I don't remember much of anything.

**Comment by Steviegt6** [comment]
Oh, yeah, we should ensure that custom trackers are fully implemented, as well.
P.S.: Did we ever decide on per-player/per-world/per-install achievements? Every single one is *doable* with the current system, and I can already envision a potential design for each.

**Comment by Solxanich** [comment]
> Is there any current reason this is considered something that would be added far in the future?

Nope, if somebody wants to pick this up sooner, they are more than welcome to. Just isn't a high priority for tML Team afaik and can come after 1.4 tML release. 



> Did we ever decide on per-player/per-world/per-install achievements

I believe the decision was implement per-world and per-install achievements and allow mods to opt-in to either. Per-player doesn't have many unique use cases, so it can be deferred in favour of the other two to start. Vanilla achievements in general could easily have been per-world.

**Comment by Steviegt6** [comment]
> 
> 
> > (This reply applies to the original post as well.)
> > Pre-made tML achievements are kind of dumb. tML is a modding API and mod loader, not a content mod in itself (save for some Patreon sets, an old April Fool's item back from when Blushie was the main dev, and a few items to help with mod compatibility (unloaded items, Starter Bag).
> 
> Keep in mind that tML is an application in and of itself. This means tML achievements (ie achievements as an application, not as modded achievements) are something to consider. For example, tML achievements may be used for steam achievements (something a modded achievement wouldn't be able to do). IE: the point with tML achievements is what could we have as Steam Achievements for tML as an application, not as a modding API. The suggestions provided match this paradigm. See Garry's Mod for example of what could be tML achievements.
> 
> Modded achievements, on the other hand, tML has no business in adding any to compare to content mods.

I still don't see why we need achievements for tML itself, though? It doesn't add anything and there's nothing that'd make sense. Nothing you do in tML screams "achievement".

**Comment by Steviegt6** [comment]
> 
> 
> > Is there any current reason this is considered something that would be added far in the future?
> 
> Nope, if somebody wants to pick this up sooner, they are more than welcome to. Just isn't a high priority for tML Team afaik and can come after 1.4 tML release.
> 
> > Did we ever decide on per-player/per-world/per-install achievements
> 
> I believe the decision was implement per-world and per-install achievements and allow mods to opt-in to either. Per-player doesn't have many unique use cases, so it can be deferred in favour of the other two to start. Vanilla achievements in general could easily have been per-world.

Per-world and per-install sounds good, but per-player should be considered too. No point in limiting a modder.
